milestones/data.py

```diff
@@ -0,0 +1,373 @@
+        course_id=unicode(course_key),
+        milestone=milestone_obj,
+        milestone_relationship_type=relationship_type,
+        active=True,
+    )
+
+
+def delete_course_milestone(course_key, milestone):
+    """
+    Removes an existing course-milestone from app/local state
+    No response currently defined for this operation
+    """
+    try:
+        internal.CourseMilestone.objects.get(
+            course_id=unicode(course_key),
+            milestone=milestone['id'],
+            active=True,
+        ).delete()
+    except internal.CourseMilestone.DoesNotExist:
+        pass
+
+
+def fetch_courses_milestones(course_keys, relationship=None, user=None):
+    """
+    Retrieves the set of milestones currently linked to the specified courses
+    Optionally pass in 'relationship' (ex. 'fulfills') to filter down the set
+    """
+    queryset = internal.CourseMilestone.objects.filter(
+        course_id__in=course_keys,
+        active=True
+    ).select_related('milestone')
+
+    # if milestones relationship type found then apply the filter
+    if relationship is not None:
+        mrt = _get_milestone_relationship_type(relationship)
+        queryset = queryset.filter(
+            milestone_relationship_type=mrt.id,
+        )
+
+    # To pull the list of milestones a user HAS, use get_user_milestones
+    # Use fetch_courses_milestones to pull the list of milestones that a user does not yet
+    # have for the specified course
+    if relationship == 'requires' and user and user.get('id', 0) > 0:
+        queryset = queryset.exclude(milestone__usermilestone__user_id=user['id'])
+
+    # Assemble the response container
+    course_milestones = []
+    if len(queryset):
+        for milestone in queryset:
+            course_milestones.append(serializers.serialize_milestone_with_course(milestone))
+
+    return course_milestones
+
+
+def create_course_content_milestone(course_key, content_key, relationship, milestone):
+    """
+    Inserts a new course-content-milestone into app/local state
+    No response currently defined for this operation
+    """
+    relationship_type = _get_milestone_relationship_type(relationship)
+    milestone_obj = serializers.deserialize_milestone(milestone)
+    internal.CourseContentMilestone.objects.get_or_create(
+        course_id=unicode(course_key),
+        content_id=unicode(content_key),
+        milestone=milestone_obj,
+        milestone_relationship_type=relationship_type,
+        active=True,
+    )
+
+
+def delete_course_content_milestone(course_key, content_key, milestone):
+    """
+    Removes an existing course-content-milestone from app/local state
+    No response currently defined for this operation
+    """
+    try:
+        internal.CourseContentMilestone.objects.get(
+            course_id=unicode(course_key),
+            content_id=unicode(content_key),
+            milestone=milestone['id'],
+            active=True,
```

Comment:
See previous note -- I've included active=True to make it easy to switch over to 'soft' deletes after we have the discussion.